[Concurrency] Look for explicit 'nonisolated' when getting isolation from protocol conformances.

simanerush opened the pull request:
If the protocol from the conformance list explicitly strips off global actor with `nonisolated`, make sure to look for this explicit isolation source when computing the resulting isolation; to prevent accidental global actor inference on the conforming type.

Resolves rdar://144798727.

simanerush committed:
[Concurrency] Look for explicit 'nonisolated' when getting isolation from protocol conformances.

simanerush commented:
@swift-ci please smoke test

AnthonyLatsis commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -5021,13 +5021,23 @@ getIsolationFromConformances(NominalTypeDecl *nominal) {
     case ActorIsolation::CallerIsolationInheriting:
     case ActorIsolation::NonisolatedUnsafe:
       break;
+    case ActorIsolation::Nonisolated:
+      if (inferredIsolation.source.kind == IsolationSource::Kind::Explicit) {
+        // We found an explicitly 'nonisolated' protocol.
+        foundIsolation = {protoIsolation,
+                          IsolationSource(proto, IsolationSource::Conformance)};
+        continue;
+      } else {
+        break;
+      }
```
Is `@MainActor` still inferred for `S` below?

```swift
nonisolated protocol P1 {}
@MainActor protocol P2: P1 {}

do {
  struct S: P2 {
    func f() {}
  }

  nonisolated func test() {
    S().f()
  }
}
```

simanerush committed:
[Concurrency] Make sure to still infer global actor if the conformance kind is explicit.

simanerush commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@swift-ci please smoke test macOS

simanerush commented:
@swift-ci please smoke test

simanerush commented:
@swift-ci please smoke test

xedin committed:
[Tests] NFC: Add a test-case for main-actor isolated struct and nonisolated protocol

xedin commented:
@swift-ci please test

hborla commented:
Is that missing a protocol requirement? `NonisolatedStruct` does not conform to any protocols.

xedin commented:
`A` does conform to `Refined`, `NonisolatedStruct` just uses for nonisolated function. We can add explicitly nonisolated global as well which gets `A` as parameter if you'd prefer.

xedin committed:
[Concurrency] SE-0449: Implied conformances to nonisolated protocols make witnesses nonisolated

Even if the requirement is stated on an isolated protocol if the
conformance is implied by a nonisolated one all of the requirements
and witnesses should be nonisolated.

xedin commented:
@swift-ci please test
